### ISSUE TITLE
WebKeys.public sometimes does not contain Scala.js assets

### ISSUE THREAD

**Issue by julienrf** [vmunier/sbt-web-scalajs#22]
Hi @vmunier,

Maybe this is not directly related to your plugin… I need to define a code generator that consumes the web assets produced by sbt-web. I have the following code in my `build.sbt`:

~~~ scala
(sourceGenerators in Compile) += Def.task {
  val assetsDirectory = (WebKeys.public in Assets).value
  // Do something with assets
}.taskValue,
~~~

Then, when I compile my project, the first time `assetsDirectory` contains only “static” assets (those that I defined in `src/main/assets`), and no Scala.js generated .js file. If I compile the project a second time, then I get the Scala.js files.

In other words, I need my code generator to run *after* the sbt-play-scalajs plugin has finished producing its assets.

Any idea of what is wrong with my approach?

**Comment by vmunier** [comment]
Hi @julienrf,

I do not see any Scala.js generated .js file at all under `assetsDirectory` when compiling, and it's strange that you see them the second time you compile.

I think you would have to use `scalaJSDev.value`/`scalaJSProd.value` to access the `-fastopt.js`/`-opt.js` file along with the sourcemap files.

**Comment by julienrf** [comment]
What I want to do is get all the assets + the generated scalajs files. With scalaJSDev.value I would only get the generated scalajs files, right?

**Comment by julienrf** [comment]
Actually I don’t understand the goal of this line: https://github.com/vmunier/sbt-play-scalajs/blob/master/src/main/scala/playscalajs/PlayScalaJS.scala#L38
If you copy the files into the assets directory, it will be included to the jar by sbt-web, no?

**Comment by julienrf** [comment]
Why don’t you instead scope your source generator to the `Assets` configuration, as described here?
https://github.com/sbt/sbt-web#writing-a-source-file-task

**Comment by julienrf** [comment]
(Sorry for the spam…)

I replaced sbt-play-scalajs by the following source generator in my `server` project:

~~~ scala
sourceGenerators in Assets += Def.task {
  Seq((fastOptJS in (client, Compile)).value.data)
}.taskValue,
~~~

And, though that does not support prod mode and sourcemaps, I think it integrates better with sbt-web. Now if I have a task that depends on `SbtWeb.public in Assets` I get all the static assets + the Scala.js output.

**Comment by vmunier** [comment]
The `scalaJSDev` part (https://github.com/vmunier/sbt-play-scalajs/blob/a0f3e1593aa0c688aadc7b6b70d0161ba4fdf808/src/main/scala/playscalajs/PlayScalaJS.scala#L37-L39) is indeed confusing.

`scalaJSDev` executes all the Scala.js tasks needed during development: 
`fastOptJS`, `packageScalaJSLauncher`, `packageJSDependencies` and collect the Scala files for sourcemap.

In the context of a Play app using `sbt-play-scalajs`, if `scalaJSDev` were added to `sourceGenerators in Assets`, the `-fastopt.js` file + Scala files used for sourcemap would be added to the final JAR when using a production command (`start`, `stage`, `dist`). I made the decision to only add the `-opt.js` file to a production archive by default, neither `-fastopt.js` nor sourcemap files are added (note that `emitSourceMaps in fullOptJS := true` can be used in a client project to add sourcemap in production).

Ideally, I think there should be a way in `sbt-web` to produce output files, which could be used during development but would be excluded from the production step.
